File: src/publisher.cpp
#include <ros/ros.h>
#include <geometry_msgs/Pose.h> // For geometry_msgs::Pose.
#include <geometry_msgs/Quaternion.h>  // For geometry_msgs::Quaternion.
#include <stdlib.h> // For rand() and RAND_MAX.
#include <tf2/LinearMath/Quaternion.h> // For quaternion calculations.
#include <tf2_geometry_msgs/tf2_geometry_msgs.h> // For transforming tf2 to message.
// #include <[PATH_TO_VISION_CODE]/[VISION_CODE].h>

int main(int argc, char **argv)
{
	// Initialize the ROS system and become a node.
	ros::init(argc, argv, "publisher_node");
	ros::NodeHandle nh;

	// Create a publisher object.
	ros::Publisher pub = nh.advertise<geometry_msgs::Pose>(
		"qr/pose", 1000); // Here we need to choose the right topic.

	// Seed the random number generator.
	srand(time(0));

	// Loop at 2Hz until the node is shut down.
	ros::Rate rate(2);
	while(ros::ok())
	{
		// Create and fill in the message randomly. The z-coordinate of 
		// the Point element defaults to zero. Only consider rotation 
		// about z-axis (yaw).
		geometry_msgs::Pose msg;
		msg.position.x = double(rand());
		msg.position.y = double(rand());
		tf2::Quaternion q;
		q.setRPY(0, 0, double(rand()));
		geometry_msgs::Quaternion q_msg = tf2::toMsg(q);
		msg.orientation = q_msg;

		// // Here fill in with the code that gets the pose of the QR code 
		// // with opencv.
		// x, y, theta = [NAMESPACE_VISION_CODE]::get_qr_pose();
		// msg.position.x = x;
		// msg.position.y = y;
		// tf2::Quaternion q;
		// q.setRPY(0, 0, theta);
		// q_msg = tf2::toMsg(q);
		// msg.orientation = q_msg;

		// Publish the message.
		pub.publish(msg);

		// Send a message to rosout with the details.
		ROS_INFO_STREAM("Publishing QR positions:"
			<< " position=" << "(" << msg.position.x << ", " << msg.position.y << ")"
			<< " orientation=" << "(" << msg.orientation.x << ", " 
			<< msg.orientation.y << ", "<< msg.orientation.z << ", " 
			<< msg.orientation.w << ")"); 

		// Wait until it's time for another iteration.
		rate.sleep();
	}
}

File: src/publisher.py
#!/usr/bin/env python3

import rospy
import random
import time
from qr_vision.msg import MarkerArray, Marker
# from [PATH_TO_VISION_CODE] import get_qr_pose
from aruco import ArucoDetector

def main():

    # Initialize the ROS system and become a node.
    rospy.init_node("publish_qr_pose", anonymous=True)

    # Create a publisher object.
    # Here we need to choose the right topic name.
    pub = rospy.Publisher('rsahackathon/poses', MarkerArray, queue_size=10)

    detector = ArucoDetector(4)

    # Loop at 2Hz until the node is shut down.
    rate = rospy.Rate(2)
    while not rospy.is_shutdown():

        # Get the positions of the markers (form: { id: (x,y,theta,time) })
        positions = detector.updatePositions()
        # print("Markers: ", positions)

        # Create the ROS message from the list of markers
        msg = MarkerArray()
        if (positions is None): continue
        for id,position in positions.items():
          marker = Marker()
          marker.name = str(id) # TODO: Add a way to give names to the markers
          marker.id = id
          marker.x = round(position[0],3)
          marker.y = round(position[1],3)
          marker.theta = round(position[2],3)
          marker.age = round(time.time() - position[3],3)
          msg.markers.append(marker)
          print("%s : (x=%s, y=%s, θ=%s) - %s" %(marker.id, marker.x, marker.y, marker.theta, marker.time))

        # Publish the message.
        pub.publish(msg)

        # Send a message to rosout with the details.
        rospy.loginfo("Publishing positions")

        # Wait until it's time for another iteration.
        rate.sleep()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass
